CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+
+### Key management
+
+Because ODIS' current OPRF function operates with the same cryptographic primitives as the [proposed
+function](https://github.com/celo-org/celo-proposals/pull/234), and because its a bad idea in
+general to reuse keys, implementation of this proposal MUST use a new set of keys, not used in any
+other service.
+
+Keys used in this service SHOULD be generated through a DKG ceremony amongst the servers that will
+operate the service. This new public key SHOULD be distributed along with client implementations and
+used to verify the values returned from the new API. Note that the existing ODIS OPRF key will
+continue to be used to verify output from the existing API.
+
+### Cryptographic interface
+
+Implementation of a domain-restricted hashing function is built on top of a partial
+[oblivious](http://www.cs.columbia.edu/~tal/4261/F19/hugo-columbia-oprf.pdf) [pseudorandom
+function](https://en.wikipedia.org/wiki/Pseudorandom_function_family) (pOPRF). It extends the
+existing OPRF, currently implemented by ODIS, by adding a second input, which is not blinded, and
+can be used by the pOPRF for rate limiting.
+
+With respect the client, this functionality has the interface $F_K(d,m)$ where $d$, $m$, and the
+output are binary strings. $K$ is a public key which commits to the output of the function, and
+allows for verification. $d$ is considered the "domain string" and is visible to the pOPRF service.
+$m$ is the message string, and is hidden from the pOPRF service. Both strings are combined
+cryptographically in a one-way function to form the output.
+
+Multiple candidate implementations exist for this functionality. Linked below is the proposed
+cryptographic design for use in the ODIS pOPRF service. Other implementation with the same interface
+may also be used to implement the following service design.
+
+<!-- TODO(victor) Replace this Notion link when the proposal is written up as a CIP -->
+[pOPRF Cryptography Proposal](https://www.notion.so/pOPRF-Cryptography-Proposal-493f1099460940f8a5d7dee4c78b4442)
+
+## Backwards Compatibility
+
+Backwards compatibility for current users of ODIS is ensured because this proposal leaves in place
+all existing functionality, including the public keys currently used to operate the service.
+
+## Security Considerations
+
+Incorrect implementation or operation of the ODIS service can have very serious effects on users of
+the service. In particular, a bug in the rate limiting code or compromise of a threshold of nodes
+could allow an attacker to bypass rate limiting and derive a user's secret from its hash or other
+commitment.
+
+Implementing this proposal involves writing new code to be included in the ODIS service, which
+implies a potential increase in the attack surface and opportunities to introduce security
+vulnerabilities. Care should be taken to ensure the correctness of the implementation, including
+testing, careful code review, and potentially third-party audits.
+
```

Comment:
Maybe an open question - would the parameters of the pOPRF vary with the domain type? For instance, would password domains use the same set of signers and threshold number of signatures as the phone number domains? Or does it make more sense to vary these parameters based on the domain (for ex, password domains may require a higher threshold of signatures for stronger security guarantees/decreased risk of collusion by a few malicious signers)?

Reply:
Varying the signer set and/or threshold based on the domain would require a distinct keys for each configuration. We don't provide a way to do this in this proposal. In this proposal the signer set is abstracted away as part of the "cryptographic functionality". If there is a need for this in the future, we could propose a further extension to ODIS.